File: 0238-product-of-array-except-self/Solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        left = [0] * n
        right = [0] * n
        left[0] = right[n-1] = 1

        # Create left multiply array
        for i in range(1, n):
            left[i] = left[i-1] * nums[i-1]
        # Create right multiply array
        for j in range(n-2, -1, -1):
            right[j] = right[j+1] * nums[j+1]
        # Create result
        for k in range(n):
            res[k] = left[k] * right[k]
        return res

File: 0238-product-of-array-except-self/Solution.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut res = vec![1; n];
        let mut left = 1;
        let mut right = 1;

        for i in 0..n {
            res[i] = left;
            left *= nums[i];
        }
        for j in (0..n).rev() {
            res[j] *= right;
            right *= nums[j];
        }

        res
    }
}